### PR TITLE
VLAN support

### PR DESCRIPTION
**Pull request by jukkar**
This set provides VLAN support for ethernet networks. Works when connecting to Linux host with both IPv4 and IPv6 and with multiple VLAN tags.
To enable VLAN support one needs to set **CONFIG_NET_VLAN=y** in config file.
To enable VLAN traffic, currently this needs to be enabled at runtime in net-shell with "_**net vlan add 100 0**_" command. This will add VLAN tag 100 to first network interface (index 0). Multiple tags can be enabled simultaneously, one just needs to set **CONFIG_NET_VLAN_COUNT=xxx** to proper value (default is 1).  The sample app will automatically enable configured VLANs.

This PR builds on top of following PRs:

* PR #6601 Checksum offloading
* PR #6076 Traffic classification

### CODE REVIEW

**Comment by pfalcon** [comment]
This apparently addresses https://github.com/zephyrproject-rtos/zephyr/issues/3234

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Nack on the current approach.

It's introducing an L2 and bearer specific attribute to the net core part. Then we have vlan_tag as a parameter everywhere, and that's really bad.
Plus, having a fixed size of net_if_config on each iface means we will use too much memory in some use case: one iface with 2 vlan, another one with 5, etc...

DAD, mcast etc... all of this don't use the vlan_tag as is. What matters here is the ip configuration only and its relationship with iface. And that's where we should address the changes.

We could split net_if in 2: 
- net_if_dev (the one that relates to an actual device):
   the same as what net_if is right now.
- net_if_virt (a "virtual" iface, which in this case would be instantiated to hold vlan ip config stuff):
  that one would hold one net_if_config and a pointer to       the net_if_dev it is attached to.

net_if being then just a generic place-holder of a union on these 2 structs, with an attribute that can be used to differentiate those.

Something like that. It's more work, but at least it won't spread vlan stuff everywhere.

The core part of vlan itself should go in L2 (either tight to ethernet, or more generic as - maybe - future tech support would require it, such as ATM based devs etc...).

**Comment by jukkar** [comment]
> Plus, having a fixed size of net_if_config on each iface means we will use too much memory in some use case: one iface with 2 vlan, another one with 5, etc...

Yep, agreed on this one. I did not try to separate IP config from net_if yet, there would have been too much changes. We could have the config set separately and only store pointer to it in net_if in order to save space when some functionality is not needed.

> 
> We could split net_if in 2:
> 
>     net_if_dev (the one that relates to an actual device):
>     the same as what net_if is right now.
>     net_if_virt (a "virtual" iface, which in this case would be instantiated to hold vlan ip config stuff):
>     that one would hold one net_if_config and a pointer to the net_if_dev it is attached to.

The problem is that currently in the code we have "struct net_if *" in many places. We could do it so that current net_if is renamed to net_if_dev, and then create new net_if that would contain IP configuration etc that is needed for the interface. We probably do not need this net_if_virt at all.

> 
> net_if being then just a generic place-holder of a union on these 2 structs, with an attribute that can be used to differentiate those.

I am not sure this is needed but lets see how this goes.

Anyway, this set enables a way to have VLAN support which was the intention. The changes in this set are quite simple actually, we just need to generalize this a bit more. The purpose of this set was to get some comments and now they are got.

I will send a new version of this set to fix various bugs, no need to review it at this time.

**Comment by jukkar** [comment]
New version uploaded that adds VLAN support only to ethernet L2 driver.
This one is still missing:
- vlan priority
- IP address optimizations (now uses too much memory)
These two are work in progress.

File: samples/net/vlan/README.rst [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+- :ref:`networking_with_qemu`
+
+Building and Running
+********************
+
+There are multiple ways to use this application. One of the most common
+usage scenario is to run vlan application inside QEMU. This is
+described in :ref:`networking_with_qemu`.
+
+Build VLAN sample application like this:
```
Comment (dbkinder) [inline samples/net/vlan/README.rst]: Follow these steps to build the VLAN sample application:

File: samples/net/vlan/README.rst [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/net/vlan
+   :board: <board to use>
+   :conf: <config file to use>
+   :goals: build
+   :compact:
+
+The default configuration file prj.conf will create two virtual lan networks
```
Comment (dbkinder) [inline samples/net/vlan/README.rst]: s/will create/creates/
s/lan/LAN/

File: samples/net/vlan/README.rst [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+with these settings:
+- VLAN tag 100: IPv4 198.51.100.1 and IPv6 2001:db8:100::1
+- VLAN tag 200: IPv4 203.0.113.1 and IPv6 2001:db8:200::1
+
+Setting up Linux Host
+=====================
+
+The :file:`samples/net/vlan/vlan-setup-linux.sh` provides a script that
+can be executed in Linux host. It creates two VLANs in Linux host and
```
Comment (dbkinder) [inline samples/net/vlan/README.rst]: s/in/on the/  (twice)

**Review by dbkinder — CHANGES_REQUESTED** [review]
suggested edits

**Comment by jukkar** [comment]
Updated new version:
- documentation fixed
- the code passes now sanitycheck
- optimization of IP addresses in network interface (can save lot of memory when having multiple network interfaces)

**Comment by jukkar** [comment]
New version that fixes sanitycheck error. I did not see this one when manually running sanitycheck but shippable run sanitycheck gives some errors.

**Comment by jukkar** [comment]
Documentation fix to satisfy sanitycheck.

**Comment by jukkar** [comment]
New version that fixes merge conflict.

File: samples/net/vlan/README.rst [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+.. zephyr-app-commands::
+   :zephyr-app: samples/net/vlan
+   :board: <board to use>
+   :conf: prj.conf
+   :goals: build
+   :compact:
+
+The default configuration file prj.conf creates two virtual LAN networks
+with these settings:
+- VLAN tag 100: IPv4 198.51.100.1 and IPv6 2001:db8:100::1
```
Comment (dbkinder) [inline samples/net/vlan/README.rst]: add a blank line before a list (otherwise the line looks like a continuation of the sentence)

**Review by dbkinder — CHANGES_REQUESTED** [review]
one last formatting fix

File: include/net/net_pkt.h [hunk 5]
```diff
@@ -128,6 +128,11 @@ struct net_pkt {
 	u8_t ieee802154_rssi; /* Received Signal Strength Indication */
 	u8_t ieee802154_lqi;  /* Link Quality Indicator */
 #endif
+
+#if defined(CONFIG_NET_VLAN)
+	u16_t vlan_tag;
+#endif /* CONFIG_NET_VLAN */
```
Comment (tbursztyka) [inline include/net/net_pkt.h]: Why net_pkt would need to store that info?

File: subsys/net/ip/net_shell.c [hunk 6]
```diff
@@ -161,6 +161,9 @@ static void iface_cb(struct net_if *iface, void *user_data)
 #if defined(CONFIG_NET_IPV6)
 	struct net_if_ipv6_prefix *prefix;
 	struct net_if_router *router;
+#endif
```
Comment (tbursztyka) [inline subsys/net/ip/net_shell.c]: don't you have a dedicated patch for shell + vlan? that part should go there.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Github was unable to load the large diffs for me...

Anyway, in arp and ethernet, could create dedicated function for setting up the ptype, hdr so it would unbloat a bit the functions. It's full of ifdefs and if/else there. It would be more readable to isolate things a bit.

File: include/net/net_ip.h [hunk 7]
```diff
@@ -349,9 +349,10 @@ static inline bool net_is_ipv6_addr_mcast(const struct in6_addr *addr)
 }
 
 struct net_if;
+struct net_if_config;
 
-extern struct net_if_addr *net_if_ipv6_addr_lookup(const struct in6_addr *addr,
-						   struct net_if **iface);
+extern struct net_if_addr *net_if_ipv6_addr_lookup(
+	const struct in6_addr *addr, struct net_if **iface);
```
[the remark this answers is not in the document]
Comment (jukkar) [reply include/net/net_ip.h]: Yes, this change is unnecessary, will fix.

File: include/net/net_pkt.h (hunk 5 above)
Comment (jukkar) [reply include/net/net_pkt.h]: I noticed that iface variable that is feed to L2 ethernet.c is not always the proper value for a specific VLAN tag. Having this value set in net_pkt is also useful as figuring it out from iface is time consuming as iface does not actually know anything about VLANs.

**Comment by jukkar** [comment]
> It's full of ifdefs and if/else there. It would be more readable to isolate things a bit.

That is unfortunate but I am not sure how to isolate things in arp.c, I will investigate if that is possible.

File: include/net/net_pkt.h (hunk 5 above)
Comment (tbursztyka) [reply include/net/net_pkt.h]: ok

**Comment by jukkar** [comment]
Updated according to comments. Also fixed couple of issues I noticed while testing VLAN support.

File: subsys/net/ip/net_shell.c (hunk 6 above)
Comment (jukkar) [reply subsys/net/ip/net_shell.c]: That patch is for creating "net vlan" command, the changes here just print vlan information when user gives "net iface" command.

**Comment by jukkar** [comment]
Updated new version as newly merged openthread code was using IP address configs that this PR changes, thus there was various compile errors.

**Comment by jukkar** [comment]
New version uploaded:

* rebased because of merge conflicts
* VLAN priority support added
* unit tests for VLAN bit field functions

**Comment by jukkar** [comment]
New version that fixes the sanity errors.

**Comment by jukkar** [comment]
Fixed the compilation if IPv4 is not enabled.

File: subsys/net/ip/net_if.c [hunk 8]
```diff
@@ -190,11 +199,12 @@ static void net_if_process_events(struct k_poll_event *event, int ev_count)
...
-			struct net_if *iface;
+			struct net_if_dev *if_dev;
+
+			if_dev = CONTAINER_OF(event->fifo, struct net_if_dev,
+					      tx_queue);
 
-			iface = CONTAINER_OF(event->fifo, struct net_if,
-					     tx_queue);
-			net_if_tx(iface);
+			net_if_foreach(iface_tx_cb, if_dev);
```
Comment (tbursztyka) [inline subsys/net/ip/net_if.c]: Ok, this is temporary as new PR will change that entirely (TC support etc...) So it won't need to loop on each iface etc...

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Merge the 3 mcux patch in one (and call the patch "Enabling VLAN in mcux" or something like that). No need to be incremental here, and maybe you want to order the patch in the list like right before the test patch maybe.

**Comment by jukkar** [comment]
Updated the PR according to comments.

**Review by dbkinder — APPROVED** [review]
thanks

**Comment by jukkar** [comment]
Updated because of changes in master. New net_if_carrier_down() func caused compile error with this PR.

**Comment by jukkar** [comment]
Uploaded because of merge conflict

**Comment by jukkar** [comment]
Added more unit tests.

**Comment by jukkar** [comment]
New version that is now on top of traffic class patches. This was needed as I could not get the tx_queue handling work reliably and the traffic class PR got rid of that part of the code in this PR.

**Comment by jukkar** [comment]
New version with sanitychecker fixes.

**Comment by jukkar** [comment]
Fixed slip support.

**Comment by jukkar** [comment]
Rebase against traffic class pr.

**Comment by jukkar** [comment]
Rebase against master. Fixed issues reported by sanitycheck.

**Comment by jukkar** [comment]
Rebase against master.

**Comment by jukkar** [comment]
* Rebased against master
* Fixed checksum_offloading tests because of VLAN changes
* Added "net iface up|down" command to net-shell

**Comment by pfalcon** [comment]
> Added "net iface up|down" command to net-shell

Would it make sense to submit that as a separate PR for quick approval? (I for example didn't yet have a chance to go thru all 32 commits of this PR.)

**Comment by jukkar** [comment]
rebase against master

**Comment by jukkar** [comment]
Removing this at this time as this seems to confuse people too much. I will re-send this, after PRs that depend on it, have been merged.

**Pull request closed without merging**